sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -495,9 +495,12 @@ object TypeCoercion {
           i
         }
 
-      case i @ In(a, b) if b.exists(_.dataType != a.dataType) =>
-        findWiderCommonType(i.children.map(_.dataType)) match {
-          case Some(finalDataType) => i.withNewChildren(i.children.map(Cast(_, finalDataType)))
+      case i @ In(value, list) if list.exists(_.dataType != value.dataType) =>
+        findWiderCommonType(list.map(_.dataType)) match {
```

Comment:
are you sure this is the behavior of binary comparison? It seems we should call `findTightestCommonType` and then `findCommonTypeForBinaryComparison`

Reply:
```
findTightestCommonType(value.dataType, listType).orElse(findCommonTypeForBinaryComparison(value.dataType, listType, conf))
```
Can not cover these cases:

- `bigint` vs `decimal`
- `float` vs `decimal`
- `double` vs `decimal`

Example:
```
-- !query 43
SELECT cast(1 as bigint) in (cast(1 as decimal(10, 0))) FROM t
-- !query 43 schema
struct<>
-- !query 43 output
org.apache.spark.sql.AnalysisException
cannot resolve '(CAST(1 AS BIGINT) IN (CAST(1 AS DECIMAL(10,0))))' due to data type mismatch: Arguments must be same type but were: bigint != decimal(10,0); line 1 pos 25

```
```
spark-sql> explain SELECT cast(1 as bigint) > (cast(1 as decimal(10, 0)));
== Physical Plan ==
*(1) Project [false AS (CAST(CAST(CAST(1 AS BIGINT) AS DECIMAL(20,0)) AS DECIMAL(20,0)) > CAST(CAST(1 AS DECIMAL(10,0)) AS DECIMAL(20,0)))#29]
+- *(1) Scan OneRowRelation[]
```

Reply:
It seems we should call `findCommonTypeForBinaryComparison` -> `findWiderTypeForDecimal` -> `findTightestCommonType`.

Because `findCommonTypeForBinaryComparison` in `PromoteStrings` in [TypeCoercion.scala#L52](https://github.com/apache/spark/blob/87ab27ef9f26274c40fbd7a597e1dc13125b8a45/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L52).  `findWiderTypeForDecimal` in `DecimalPrecision` in [TypeCoercion.scala#L53](https://github.com/apache/spark/blob/87ab27ef9f26274c40fbd7a597e1dc13125b8a45/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L53). `findTightestCommonType ` in `ImplicitTypeCasts` in [TypeCoercion.scala#L63](https://github.com/apache/spark/blob/87ab27ef9f26274c40fbd7a597e1dc13125b8a45/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L63).

Reply:
After taking a closer look, the existing logic for binary comparison seems to be:
```
if (not decimal type) {
  apply findTightestCommonType
  apply findCommonTypeForBinaryComparison
} else {
  if (both sides are decimal) {
    apply widerDecimalType 
  } if (one side is not decimal) {
    apply function integralAndDecimalLiteral
    apply function nondecimalAndDecimal
  }
}
```

Reply:
@cloud-fan @maropu It seems the logic is `PromoteStrings` -> `DecimalPrecision` -> `ImplicitTypeCasts`:
```scala
scala> spark.conf.set("spark.sql.optimizer.planChangeLog.level", "WARN")

scala> spark.sql("select  bigint(1) > double(2), 1 > '1', bigint(1) > cast(1 as decimal(10, 0)) ").explain
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1: Batch Hints has no effect.
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1: Batch Simple Sanity Check has no effect.
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1: Batch Substitution has no effect.
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions ===
!'Project [unresolvedalias(('bigint(1) > 'double(2)), None), unresolvedalias((1 > 1), None), unresolvedalias(('bigint(1) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                             +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.ResolveTimeZone ===
 'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                                                  +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.TypeCoercion$PromoteStrings ===
!'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                                                  +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.DecimalPrecision ===
!'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(cast(1 as bigint) as decimal(20,0)) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                                                               +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.TypeCoercion$ImplicitTypeCasts ===
!'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(cast(1 as bigint) as decimal(20,0)) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(cast(1 as bigint) as double) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(cast(1 as bigint) as decimal(20,0)) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelationscala> spark.conf.set("spark.sql.optimizer.planChangeLog.level", "WARN")

scala> spark.sql("select  bigint(1) > double(2), 1 > '1', bigint(1) > cast(1 as decimal(10, 0)) ").explain
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1: Batch Hints has no effect.
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1: Batch Simple Sanity Check has no effect.
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1: Batch Substitution has no effect.
20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions ===
!'Project [unresolvedalias(('bigint(1) > 'double(2)), None), unresolvedalias((1 > 1), None), unresolvedalias(('bigint(1) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                             +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.ResolveTimeZone ===
 'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                                                  +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.TypeCoercion$PromoteStrings ===
!'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > 1), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                                                  +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.DecimalPrecision ===
!'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(1 as bigint) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(cast(1 as bigint) as decimal(20,0)) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation                                                                                                                                                                               +- OneRowRelation

20/02/02 19:31:10 WARN HiveSessionStateBuilder$$anon$1:
=== Applying Rule org.apache.spark.sql.catalyst.analysis.TypeCoercion$ImplicitTypeCasts ===
!'Project [unresolvedalias((cast(1 as bigint) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(cast(1 as bigint) as decimal(20,0)) > cast(1 as decimal(10,0))), None)]   'Project [unresolvedalias((cast(cast(1 as bigint) as double) > cast(2 as double)), None), unresolvedalias((1 > cast(1 as int)), None), unresolvedalias((cast(cast(1 as bigint) as decimal(20,0)) > cast(1 as decimal(10,0))), None)]
 +- OneRowRelation
```